src/Aspire.Hosting.Analyzers/Infrastructure/WellKnownTypes.cs

```diff
@@ -0,0 +1,157 @@
+
+    [Conditional("DEBUG")]
+    private static void AssertEnumAndTableInSync()
+    {
+        for (var i = 0; i < WellKnownTypeData.WellKnownTypeNames.Length; i++)
+        {
+            var name = WellKnownTypeData.WellKnownTypeNames[i];
+            var typeId = (WellKnownTypeData.WellKnownType)i;
+
+            var typeIdName = typeId.ToString().Replace("__", "+").Replace('_', '.');
+
+            var separator = name.IndexOf('`');
+            if (separator >= 0)
+            {
+                // Ignore type parameter qualifier for generic types.
+                name = name.Substring(0, separator);
+                typeIdName = typeIdName.Substring(0, separator);
+            }
+
+            Debug.Assert(name == typeIdName, $"Enum name ({typeIdName}) and type name ({name}) must match at {i}");
+        }
+    }
+
+    private WellKnownTypes(Compilation compilation)
+    {
+        _lazyWellKnownTypes = new INamedTypeSymbol?[WellKnownTypeData.WellKnownTypeNames.Length];
+        _compilation = compilation;
+    }
+
+    public INamedTypeSymbol Get(SpecialType type)
+    {
+        return _compilation.GetSpecialType(type);
+    }
+
+    public INamedTypeSymbol Get(WellKnownTypeData.WellKnownType type)
+    {
+        var index = (int)type;
+        var symbol = _lazyWellKnownTypes[index];
+        if (symbol is not null)
+        {
+            return symbol;
+        }
+
+        // Symbol hasn't been added to the cache yet.
+        // Resolve symbol from name, cache, and return.
+        return GetAndCache(index);
+    }
+
+    private INamedTypeSymbol GetAndCache(int index)
+    {
+        var result = GetTypeByMetadataNameInTargetAssembly(WellKnownTypeData.WellKnownTypeNames[index])
+            ?? throw new InvalidOperationException($"Failed to resolve well-known type '{WellKnownTypeData.WellKnownTypeNames[index]}'.");
+
+        Interlocked.CompareExchange(ref _lazyWellKnownTypes[index], result, null);
+
+        // GetTypeByMetadataName should always return the same instance for a name.
+        // To ensure we have a consistent value, for thread safety, return symbol set in the array.
+        return _lazyWellKnownTypes[index]!;
+    }
+
+    // Filter for types within well-known (framework-owned) assemblies only.
+    private INamedTypeSymbol? GetTypeByMetadataNameInTargetAssembly(string metadataName)
+    {
+        var types = _compilation.GetTypesByMetadataName(metadataName);
+        if (types.Length == 0)
+        {
+            return null;
+        }
+
+        if (types.Length == 1)
+        {
+            return types[0];
+        }
+
+        // Multiple types match the name. This is most likely caused by someone reusing the namespace + type name in their apps or libraries.
+        // Workaround this situation by prioritizing types in System and Microsoft assemblies.
+        foreach (var type in types)
+        {
+            if (type.ContainingAssembly.Identity.Name.StartsWith("System.", StringComparison.Ordinal)
```

Comment:
We had to apply this workaround in the ASP.NET Core repo because the `Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions` type name was used both by the shared framework and a third-party dependency in some cases. This would cause the `GetTypeByMetadataName` to throw. See https://github.com/dotnet/aspnetcore/issues/50836#issuecomment-1813301014 for more info.

You might be able to get away with removing this workaround given the set of types in your WellKnownTypes cache.